pkg/backend/path_creds.go

```diff
@@ -68,23 +68,28 @@ func (b *backend) credsUpdateOperation(ctx context.Context, req *logical.Request
 
 	key := credKey(data.Get("name").(string))
 
-	code, ok := data.GetOk("code")
-	if !ok {
-		return logical.ErrorResponse("missing code"), nil
+	var tok *oauth2.Token
+
+	if code, ok := data.GetOk("code"); ok {
+		cb := p.NewExchangeConfigBuilder(c.ClientID, c.ClientSecret)
+
+		if redirectURL, ok := data.GetOk("redirect_url"); ok {
+			cb = cb.WithRedirectURL(redirectURL.(string))
+		}
+
+		tok, err = cb.Build().Exchange(ctx, code.(string))
+		if _, ok := err.(*oauth2.RetrieveError); ok {
+			return logical.ErrorResponse("invalid code"), nil
+		} else if err != nil {
+			return nil, err
+		}
+	} else if refreshToken, ok := data.GetOk("refresh_token"); ok {
+		tok = &oauth2.Token{RefreshToken: refreshToken.(string)}
+	} else {
+		return logical.ErrorResponse("missing code or refresh_token"), nil
 	}
 
-	cb := p.NewExchangeConfigBuilder(c.ClientID, c.ClientSecret)
-
-	if redirectURL, ok := data.GetOk("redirect_url"); ok {
-		cb = cb.WithRedirectURL(redirectURL.(string))
-	}
-
-	tok, err := cb.Build().Exchange(ctx, code.(string))
-	if _, ok := err.(*oauth2.RetrieveError); ok {
-		return logical.ErrorResponse("invalid code"), nil
-	} else if err != nil {
-		return nil, err
-	}
+	b.logger.Debug("creds update", "token", tok)
```

Comment:
Does the token have a `Stringer` implementation? If not, I worry that this would leak token information to a debug log; we should avoid that.

Reply:
The token information does go into the log so you're right, I should probably delete that.  It had occurred to me.  vault-plugin-auth-jwt has a configuration option of whether or not to put token information into debug logs, default off.  I will just remove it.